File: src/python/stfs_extract.py
def stfs_extract(stfs_file_path: str, dest_path: str) -> str:
  """
  Extract all files from a CON file.
  
  Parameters
  ----------
  stfs_file_path : str
    The path of the CON file to be extracted.
  dest_path : str
    The folder path where you want the files to be extracted to.
  """
  con = STFS(stfs_file_path)
  
  # Create directories
  for filename in con.allfiles:
    if con.allfiles[filename].isdirectory:
      dirpath = filename[1:]
      dircomponents = dirpath.split("/")
      for i in range(len(dircomponents)):
        dir_path = "\\".join(dircomponents[:i+1])
        new_folder_path = f"{dest_path}\\{dir_path}"
        try:
          os.mkdir(new_folder_path)
        except OSError:
          pass
        
  # Writing files
  for filename in con.allfiles:
    if filename == "/songs/":
      continue
    if not con.allfiles[filename].isdirectory:
      
      file_bytes = con.read_file(con.allfiles[filename])
      new_file_path = f"{dest_path}{filename}"
      open(new_file_path, "wb").write(file_bytes)
  
if __name__ == '__main__':
  import argparse
  import os
  from lib.stfs import STFS
  
  parser = argparse.ArgumentParser(description='RBToolsJS: CON Extractor CLI', epilog='By Ruggery Iury Corrêa.')
  parser.add_argument('stfs_file_path', help='The RB3CON file you want to extract and print its contents', type=str)
  parser.add_argument('dest_path', help='The folder path where you want the files to be extracted to', type=str)

  arg = parser.parse_args()
  
  stfs_extract(arg.stfs_file_path, arg.dest_path)

File: src/lib/python/stfsExtract.ts
import { DirPath, execAsync, FilePath, type FilePathLikeTypes } from 'node-lib'
import { pathLikeToString } from 'node-lib'
import { PythonExecutionError } from '../../errors'
import { RBTools } from '../../index'

/**
 * Python script: Asynchronously extracts the CON file contents and returns the folder path where all contents were extracted.
 * - - - -
 * @param {FilePathLikeTypes} stfsFilePath The path of the CON file.
 * @param {FilePathLikeTypes} destPath The folder path where you want the files to be extracted to.
 * @returns {Promise<DirPath>}
 */
export const stfsExtract = async (stfsFilePath: FilePathLikeTypes, destPath: FilePathLikeTypes): Promise<DirPath> => {
  const moduleName = 'stfs_extract.py'
  const pyPath = FilePath.of(RBTools.python.path, moduleName)
  const src = FilePath.of(pathLikeToString(stfsFilePath))
  const dest = DirPath.of(pathLikeToString(destPath))

  if (dest.exists) await dest.deleteDir()
  await dest.mkDir()

  const command = `python ${moduleName} "${src.path}" "${dest.path}"`
  const { stderr } = await execAsync(command, { windowsHide: true, cwd: pyPath.root })
  if (stderr) throw new PythonExecutionError(stderr)

  return dest
}

/**
 * Python script: Asynchronously extracts all files from a CON file on the root directory of the
 * destination path and returns the folder path where all contents were extracted.
 * - - - -
 * @param {FilePathLikeTypes} stfsFilePath The path of the CON file.
 * @param {FilePathLikeTypes} destPath The folder path where you want the files to be extracted to.
 * @returns {Promise<DirPath>}
 */
export const stfsExtractAllFiles = async (stfsFilePath: FilePathLikeTypes, destPath: FilePathLikeTypes): Promise<DirPath> => {
  const moduleName = 'stfs_extract_all_files.py'
  const pyPath = FilePath.of(RBTools.python.path, moduleName)
  const src = FilePath.of(pathLikeToString(stfsFilePath))
  const dest = DirPath.of(pathLikeToString(destPath))

  if (dest.exists) await dest.deleteDir()
  await dest.mkDir()

  const command = `python ${moduleName} "${src.path}" "${dest.path}"`
  const { stderr } = await execAsync(command, { windowsHide: true, cwd: pyPath.root })
  if (stderr) throw new PythonExecutionError(stderr)

  return dest
}
